### PR TITLE
Improve draw performance of `Waveform`

### PR DESCRIPTION
**Pull request by EVAST9919**
Colours in waveform aren't being stored anywhere and recomputed every frame for each quad. While `Interpolation.ValueAt` uses a *correct* way of interpolating between colours by first converting them to linear and operation result to srgb, there's an underlying power function, which in combination with amount of quads and total interpolations results in draw performance loss.
Снимок экрана 2024-05-14 192528
I chose just to avoid space conversion as a whole, and while resulting colour looks a bit darker, I think performance gain worth it.

|master|cc3cb43d8f1c975d15848695b076d4fd51f3d2f5|f8043ea8eb18d279efc82fc7de8b874537da0d69|
|---|---|---|
|osu-waveform-master|osu-waveform-pr|osu-waveform-no-interpolation|
|waveform-master|waveform-pr|waveform-no-interpolation|

cc3cb43d8f1c975d15848695b076d4fd51f3d2f5 may also help a little in other scenarios. For exapmple computing child colour (which is white by default in most cases) while applying colour to the parent (or changing parent's alpha) wont trigger colour conversion:
|master|pr|
|---|---|
|master|pr|

### CODE REVIEW

**Comment by smoogipoo** [comment]
Should we be doing gamma-correct interpolation on `Color4`s in the first place? Part of me feels like that's wrong because we don't usually assume `Color4` has a colour space until it's resolved into an `SRGBColour`? I'm curious to know @bdach and @peppy's opinions.

---

One could think of it as comparing to shader/gradient interpolation, where currently it behaves like this:

https://github.com/ppy/osu-framework/assets/1329837/3872a74c-daba-4f32-84bc-39023aa18a8a

My proposal would be:
```diff
diff --git a/osu.Framework/Utils/Interpolation.cs b/osu.Framework/Utils/Interpolation.cs
index 68b077bf0..e01ce050f 100644
--- a/osu.Framework/Utils/Interpolation.cs
+++ b/osu.Framework/Utils/Interpolation.cs
@@ -247,7 +247,7 @@ public static EdgeEffectParameters ValueAt(double time, EdgeEffectParameters sta
                 };

             public static SRGBColour ValueAt(double time, SRGBColour startColour, SRGBColour endColour, double startTime, double endTime, in TEasing easing)
-                => ValueAt(time, (Color4)startColour, (Color4)endColour, startTime, endTime, easing);
+                => ValueAt(time, startColour.Linear, endColour.Linear, startTime, endTime, easing).ToSRGB();

             /// <summary>
             /// Interpolates between two sRGB <see cref="Color4"/>s in a linear (gamma-correct) RGB space.
@@ -266,8 +266,8 @@ public static Color4 ValueAt(double time, Color4 startColour, Color4 endColour,
                 if (duration == 0 || current == 0)
                     return startColour;

-                var startLinear = startColour.ToLinear();
-                var endLinear = endColour.ToLinear();
+                var startLinear = startColour;
+                var endLinear = endColour;

                 float t = Math.Max(0, Math.Min(1, (float)easing.ApplyEasing(current / duration)));

@@ -275,7 +275,7 @@ public static Color4 ValueAt(double time, Color4 startColour, Color4 endColour,
                     startLinear.R + t * (endLinear.R - startLinear.R),
                     startLinear.G + t * (endLinear.G - startLinear.G),
                     startLinear.B + t * (endLinear.B - startLinear.B),
-                    startLinear.A + t * (endLinear.A - startLinear.A)).ToSRGB();
+                    startLinear.A + t * (endLinear.A - startLinear.A));
             }

             public static Colour4 ValueAt(double time, Colour4 startColour, Colour4 endColour, double startTime, double endTime, in TEasing easing)
```

Which allows gamma-correct interpolation to happen via `SRGBColour`, resulting in:

https://github.com/ppy/osu-framework/assets/1329837/423ebdc2-6904-4849-b967-f0a10b65c211

Test:
```csharp
// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using osu.Framework.Graphics;
using osu.Framework.Graphics.Colour;
using osu.Framework.Graphics.Containers;
using osu.Framework.Graphics.Shapes;
using osu.Framework.Input.Events;
using osu.Framework.Utils;
using osuTK;
using osuTK.Graphics;

namespace osu.Framework.Tests.Visual
{
    public partial class TestSceneTesting : FrameworkTestScene
    {
        public TestSceneTesting()
        {
            Child = new Container
            {
                Anchor = Anchor.Centre,
                Origin = Anchor.Centre,
                Size = new Vector2(500, 50),
                Children = new Drawable[]
                {
                    new Box
                    {
                        RelativeSizeAxes = Axes.Both,
                        Colour = ColourInfo.GradientHorizontal(Color4.Red, Color4.Blue)
                    },
                    new InterpolatingColouredBox
                    {
                        Origin = Anchor.Centre,
                        Anchor = Anchor.CentreLeft,
                        Size = new Vector2(10, 60)
                    }
                }
            };
        }

        private partial class InterpolatingColouredBox : Box
        {
            protected override void Update()
            {
                base.Update();
                Colour = Interpolation.ValueAt(Position.X / Parent!.DrawWidth, Color4.Red, Color4.Blue, 0, 1);
            }

            protected override bool OnDragStart(DragStartEvent e) => true;

            protected override void OnDrag(DragEvent e)
            {
                Position = new Vector2(e.MousePosition.X, 0);
            }
        }
    }
}
```

---

Another question I want to ask here, is whether we want to do gamma correct interpolation _at all_. That is, perhaps my diff above is still not doing it how we really ought to be doing it but instead requiring people to pass colours via `ToLinear()` and `ToSRGB()` themselves to achieve this same result.

The reason I say that is because `ColourInfo` is already doing gamma **incorrect** interpolation:

https://github.com/ppy/osu-framework/blob/812e5290b88142271040fb4114c7e0d5c4ac2975/osu.Framework/Graphics/Colour/ColourInfo.cs#L94-L96

---

Yes, I'm aware this would be a breaking change.

**Comment by EVAST9919** [comment]
I remember there was a debate about srgb like half a year ago and @frenzibyte has been working in this area, so tagging him as well.

**Comment by frenzibyte** [comment]
Last time we discussed this we settled on the following:
 - All `Color4`s are considered to be in "sRGB space" and are converted to linear space only for performing "gamma-correct" interpolations.
 - `ColourInfo.Gradient{Horizontal,Vertical}` should interpolate colours in gamma-incorrect space, to match rendering of gradients in figma.
 - `Interpolation.ValueAt` should interpolate colours in gamma-correct space, to match general colour interpolation functions like `d3.interpolateRgb.gamma(2.2)` which is already used in web to pick the star difficulty colour (see https://github.com/ppy/osu/pull/14119#discussion_r682129668).

This is all covered in a test scene we have framework-side (`TestSceneGammaCorrection`):
 - master:
   CleanShot 2024-05-15 at 02 56 56

 - your proposal:
   CleanShot 2024-05-15 at 02 48 34

In summary, whether we should do gamma-correct or gamma-incorrect interpolation depends on the reference design we're basing our work on. For this case, I would say make it gamma-incorrect to match how we (and figma) render gradients, as we use gamma-incorrect interpolation for that (look at the `ColourInfo.GradientHorizontal` case in the first screenshot above).

**Review by frenzibyte — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
@frenzibyte You're misunderstanding my point. The question is whether we _expect_ to do gamma correct interpolation in _all cases_ by _default_.

CSS does not do this by default, and neither does most software - please read https://erikmcclure.com/blog/everyone-does-srgb-wrong-because/. Here's an example where CSS has to interpolate between colours without using a gradient:
```html
<html>
    <head>
        <style>
            .box {
                width: 100px;
                height: 100px;
                background-color: red;
                animation: colourFade 2s linear infinite;
            }

            @keyframes colourFade {
                0% { background-color: red; }
                100% { background-color: blue; }
            }
        </style>
    </head>
    <body>
        <div class="box"></div>
    </body>
</html>
```

> All `Color4`s are considered to be in "sRGB space" and are converted to linear space only for performing "gamma-correct" interpolations.

I need confirmation of this from @peppy and @bdach

> `Interpolation.ValueAt` should interpolate colours in gamma-correct space, to match general colour interpolation functions like `d3.interpolateRgb.gamma(2.2)` which is already used in web to pick the star difficulty colour

However if you'll notice, `d3.interpolateRgb(a, b)` is also a thing. `.gamma(2.2)(a, b)` is an _explicit_ functionality added on top. Right now, the _implicit_ functionality we have is gamma-correct.

---

Again, yes I'm aware this is a breaking change. I'm aware of that test scene and the implications therein.

**Comment by frenzibyte** [comment]
I see what you're getting at. My only concern by then is that we still have to do gamma-correct interpolation for picking the star difficulty colour from the spectrum, by defining an explicit method for gamma-correct interpolation or otherwise.

**Comment by smoogipoo** [comment]
Sure, and may suggestion there is to resolve it into an `SRGBColour` to understand the user's intention. When `SRGBColour` is used, it's assumed the user wants to do gamma-correct interpolation. When `Color4` is used, no colour spaces are assumed so the result is _possibly_ (but not always!) gamma-incorrect.

As I mentioned at the very beginning, this all hinges on whether `Color4` is assumed to have a colour space or not. My understanding is that it's not, which contradicts your statement, and I need confirmation of that.

File: osu.Framework/Graphics/Audio/WaveformGraph.cs [hunk 1]
```diff
@@ -377,6 +376,17 @@ protected override void Draw(IRenderer renderer)
                 shader.Unbind();
             }
 
+            // Ideally we would want to use Interpolation.ValueAt
+            // however converting the colour between colour spaces multiple times for all the quads is expensive.
+            private static Color4 linearInterpolate(Color4 first, Color4 second, float value)
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Graphics/Audio/WaveformGraph.cs:381]: You understand that that's the entire point of this PR, right? That @EVAST9919 changed it to use gamma-incorrect strictly because `Interpolation.ValueAt()` assumes sRGB and to wants to do the right thing.

`SRGBColour.Interpolation` doesn't exist. If you mean `ColourInfo.Interpolate` then that's also going to be doing gamma-incorrect but it's also interpolating over a quad which is not what we want here.

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Framework/Graphics/Audio/WaveformGraph.cs:381]: My comment strayed away from what I meant, I was simply referring to the name of the method, it says "linear interpolation", which is...not the case? Depends on what that means, it reads "interpolation in linear space" to me.

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Graphics/Audio/WaveformGraph.cs:381]: I read it as "linearly interpolate between two colours".

**Comment by bdach** [comment]
> > All `Color4`s are considered to be in "sRGB space" and are converted to linear space only for performing "gamma-correct" interpolations.
> 
> I need confirmation of this

That is what I remember the conclusion being the last time this ~~fiasco~~ topic came up, yeah. I'd have to go look at the conversations again to be 100% sure because this stuff is cursed, but against my recall alone that does track.

**Comment by smoogipoo** [comment]
Since it seems that the consensus of at least >50% is that we expect interpolation to be gamma correct, let's close this PR as not what we want and take the performance hit.

**Comment by EVAST9919** [comment]
Is cc3cb43d8f1c975d15848695b076d4fd51f3d2f5 worth a pr?

**Comment by smoogipoo** [comment]
Sure, but make sure you benchmark it first to see whether anything is actually gained.
